File: leetcode3/이진희/476. Number Complement.java
/*

1. 아이디어 : 10진법을 이진법으로 바꾸고, 1과 0을 바꾼 후, 값을 계산하는 문제
              10진법이 포함되는 이진법의 범위를 파악후, 원래 값의 num을 빼면 된다

2. 시간복잡도 : O(31)

3. 자료구조/알고리즘 : 비트 연산

 */

class Solution {
    public int findComplement(int num) {
        int sum = 1;
        int idx = 1;
        while(num>sum) {
            sum += idx*2;
            idx*=2;
        }

        return sum - num;
    }
}

File: leetcode3/변지협/v2/476. Number Complement.py

'''
1. 아이디어 :
이진수 구하고 1을 0으로, 0을 1로 바꾼 후 다시 정수로 변환한다.

2. 시간복잡도 :
o(n)

3. 자료구조/알고리즘 :
'''

class Solution(object):
    def getBinary(self, n):
        s = ''
        while True:
            if n == 0:
                break
            tmp = n % 2
            n = n // 2
            s += str(tmp)

        return s[::-1]

    def getInt(self, b):
        num = 0
        b = b[::-1]
        n = len(b)
        # print(b)
        for i in range(n):
            if int(b[i]) %2 != 0:
                num +=2 ** i

        return num

    def findComplement(self, num):
        """
        :type num: int
        :rtype: int
        """
        # print(self.getBinary(num))
        b = self.getBinary(num)
        tmp = ''
        for i in b:
            tmp += '0' if i == '1' else '1'

        return self.getInt(tmp)        